components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,398 @@
+        to_frozen_array(self.effects.as_slice(), cx)
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-playeffect>
+    fn PlayEffect(
+        &self,
+        type_: GamepadHapticEffectType,
+        params: &GamepadEffectParameters,
+    ) -> Rc<Promise> {
+        let in_realm_proof = AlreadyInRealm::assert();
+        let playing_effect_promise =
+            Promise::new_in_current_realm(InRealm::Already(&in_realm_proof));
+
+        // <https://www.w3.org/TR/gamepad/#dfn-valid-effect>
+        match type_ {
+            // <https://www.w3.org/TR/gamepad/#dfn-valid-dual-rumble-effect>
+            GamepadHapticEffectType::Dual_rumble => {
+                if *params.strongMagnitude < 0.0 || *params.strongMagnitude > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Strong magnitude value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.weakMagnitude < 0.0 || *params.weakMagnitude > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Weak magnitude value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                }
+            },
+            // <https://www.w3.org/TR/gamepad/#dfn-valid-trigger-rumble-effect>
+            GamepadHapticEffectType::Trigger_rumble => {
+                if *params.strongMagnitude < 0.0 || *params.strongMagnitude > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Strong magnitude value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.weakMagnitude < 0.0 || *params.weakMagnitude > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Weak magnitude value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.leftTrigger < 0.0 || *params.leftTrigger > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Left trigger value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                } else if *params.rightTrigger < 0.0 || *params.rightTrigger > 1.0 {
+                    playing_effect_promise.reject_error(Error::Type(
+                        "Right trigger value is not within range of 0.0 to 1.0.".to_string(),
+                    ));
+                    return playing_effect_promise;
+                }
+            },
+        }
+
+        let document = self.global().as_window().Document();
+        if !document.is_fully_active() {
```

Comment:
Now that https://github.com/servo/servo/pull/32635 has merged, you can try adding the "or document's [visibility state](https://html.spec.whatwg.org/multipage/interaction.html#visibility-state) is "hidden"" part.